### DISCUSSION TITLE
How to force CMake to use the copy of vcpkg included as a submodule for a project?

### DISCUSSION THREAD

**Discussion by crgnam** [microsoft/vcpkg#38332] [Q&A] [answered]
I have a personal copy of `vcpkg` installed on my machine however I have a project that includes `vcpkg` as a submodule.  Because it has a different baseline, I would like CMake to find and use the vcpkg submodule.  I have tried:

```
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake" CACHE FILEPATH "Vcpkg toolchain file")
```

however this does not work, and upon running the CMake configuration (with Visual Studio) it shows:

```
1> CMake generation started for configuration: 'x64-Debug'.
1> Found and using vcpkg toolchain file (<MY_USER_DIR>/vcpkg/scripts/buildsystems/vcpkg.cmake).
```

I have tried deleting CMakeChache.  I have noticed if I just completely delete my peronal copy of `vcpkg`, it correctly finds the submodule version.  But whenever another copy is present on the machine, it seems to find that first.

How can I force it to default to the vcpkg submodule included in the repository?  (And possibly, allow a user to override this should they *want* to use their copy of vcpkg)

**Comment by crgnam** [comment]
I can of course modify the `CMakeSettings.json` to directly use the correct `CMAKE_TOOLCHAIN_FILE` however this is cumberome as this file is not tracked in the repository, and I would like other users of the project to not need to change those sorts of settings for this.  I am struggling to see why:

```
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake" CACHE FILEPATH "Vcpkg toolchain file")
```

is not working.  It seems to have no effect at all.

**Reply by dg0yt** [reply, marked as answer]
Cache variables are input variables. If you want to override user input, you must either use the `INTERNAL` type, the `FORCE` keyword, or a normal variable.
And of course it must come before the  `project` command.

**Reply by crgnam** [reply]
Thanks this resolved it!

For others who may stumble upon this, here is the relevant part of my solution:

root `CMakeLists.txt`:
```
...
if("${CMAKE_TOOLCHAIN_FILE}" STREQUAL "")
    set(MY_VCPKG_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake")
    set(CMAKE_TOOLCHAIN_FILE "${MY_VCPKG_TOOLCHAIN_FILE}" 
        CACHE FILEPATH "Vcpkg toolchain file" FORCE)
    ...
endif()
...
project(...
```

This allows it to default to the included vcpkg submodule if nothing is specified.  It also allows users to either manually specify `-DCMAKE_TOOLCHAIN_FILE=OFF` or `-DCMAKE_TOOLCHAIN_FILE=<path/to/vcpkg>/vcpkg.cmake` if they so choose.  I believe this would cause issues if they wanted to use a separate vcpkg toolchain file, but that is not currently a concern for me, and I believe it could be addressed later.
